xcm/xcm-builder/src/fungibles_adapter.rs

```diff
@@ -0,0 +1,257 @@
+		let prefix = Prefix::get();
+		let id = id.borrow();
+		if !prefix.iter().enumerate().all(|(index, item)| id.at(index) == Some(item)) { return Err(()) }
+		match id.at(prefix.len()) {
+			Some(Junction::GeneralIndex { id }) => ConvertAssetId::convert_ref(id),
+			_ => Err(()),
+		}
+	}
+	fn reverse_ref(what: impl Borrow<AssetId>) -> result::Result<MultiLocation, ()> {
+		let mut location = Prefix::get();
+		let id = ConvertAssetId::reverse_ref(what)?;
+		location.push(Junction::GeneralIndex { id }).map_err(|_| ())?;
+		Ok(location)
+	}
+}
+
+pub trait MatchesFungibles<AssetId, Balance> {
+	fn matches_fungibles(a: &MultiAsset) -> result::Result<(AssetId, Balance), Error>;
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(30)]
+impl<
+	AssetId: Clone,
+	Balance: Clone,
+> MatchesFungibles<AssetId, Balance> for Tuple {
+	fn matches_fungibles(a: &MultiAsset) -> result::Result<(AssetId, Balance), Error> {
+		for_tuples!( #(
+			match Tuple::matches_fungibles(a) { o @ Ok(_) => return o, _ => () }
+		)* );
+		Err(Error::AssetNotFound)
+	}
+}
+
+pub struct ConvertedConcreteAssetId<AssetId, Balance, ConvertAssetId, ConvertBalance>(
+	PhantomData<(AssetId, Balance, ConvertAssetId, ConvertBalance)>
+);
+impl<
+	AssetId: Clone,
+	Balance: Clone,
+	ConvertAssetId: Convert<MultiLocation, AssetId>,
+	ConvertBalance: Convert<u128, Balance>,
+> MatchesFungibles<AssetId, Balance> for
+	ConvertedConcreteAssetId<AssetId, Balance, ConvertAssetId, ConvertBalance>
+{
+	fn matches_fungibles(a: &MultiAsset) -> result::Result<(AssetId, Balance), Error> {
+		let (id, amount) = match a {
+			MultiAsset::ConcreteFungible { id, amount } => (id, amount),
+			_ => return Err(Error::AssetNotFound),
+		};
+		let what = ConvertAssetId::convert_ref(id).map_err(|_| Error::AssetIdConversionFailed)?;
+		let amount = ConvertBalance::convert_ref(amount).map_err(|_| Error::AmountToBalanceConversionFailed)?;
+		Ok((what, amount))
+	}
+}
+
+pub struct ConvertedAbstractAssetId<AssetId, Balance, ConvertAssetId, ConvertBalance>(
+	PhantomData<(AssetId, Balance, ConvertAssetId, ConvertBalance)>
+);
+impl<
+	AssetId: Clone,
+	Balance: Clone,
+	ConvertAssetId: Convert<Vec<u8>, AssetId>,
+	ConvertBalance: Convert<u128, Balance>,
+> MatchesFungibles<AssetId, Balance> for
+	ConvertedAbstractAssetId<AssetId, Balance, ConvertAssetId, ConvertBalance>
+{
+	fn matches_fungibles(a: &MultiAsset) -> result::Result<(AssetId, Balance), Error> {
+		let (id, amount) = match a {
+			MultiAsset::AbstractFungible { id, amount } => (id, amount),
+			_ => return Err(Error::AssetNotFound),
+		};
+		let what = ConvertAssetId::convert_ref(id).map_err(|_| Error::AssetIdConversionFailed)?;
+		let amount = ConvertBalance::convert_ref(amount).map_err(|_| Error::AmountToBalanceConversionFailed)?;
+		Ok((what, amount))
+	}
+}
+
+pub struct FungiblesAdapter<Assets, Matcher, AccountIdConverter, AccountId>(
+	PhantomData<(Assets, Matcher, AccountIdConverter, AccountId)>
+);
+
+impl<
+	Assets: Fungibles<AccountId>,
```

Comment:
Only assets `deposit` and `withdraw` is used in in this impl, I think we could use a minimum trait bound here instead including only these two fn. In this way `FungiblesAdapter` would be easier to be reused with other assets impl like `orml-tokens`, without implementing all methods in `Fungibles`.

Reply:
with the possible exception of `transfer`, all will be needed for XCM eventually.